speechbrain/utils/distributed.py

```diff
@@ -275,6 +274,13 @@ def ddp_init_group(run_opts):
             + " communication protocol doesn't exist."
         )
 
+    # Switch to the right context ; this is needed for DDP barriers
+    # otherwise, everything before an instance of `Brain(...)`
+    # is going to have `LOCAL_RANK` equal to 0.
```

Comment:
The local rank is set by torchrun, not brain. Isn't this instead about whether the device is set?

Reply:
I know this is not set by Brain but instead by torchrun (btw, one thing we could do is to remove torchrun from SB and directly dynamically set all the flags etc something like pytorch lightning does).

I was maybe unclear but I meant the device retrieved by torch. When you are doing `torch.cuda.current_device()` you'll get a different value (always 0) than the LOCAL_RANK. After creating a Brain instance (hence setting torch.cuda.device), you'll get the cuda device to be LOCAL_RANK. I believe this is the reason why we had troubles. 

Behind the hood torch might be using the cuda device and not the LOCAL_RANK for the barriers. (this is the most likely hypothesis given my experiments that showed that setting torch.device to the LOCAL_RANK was solving our issues).

Reply:
Again this seems to be a weird choice by the PyTorch team but looks like you did proper research and this is the right approach.